toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+	snapshot          = app.Flag("snapshot", "Path to the rpm snapshot .json file.").ExistingFile()
+	outputSummaryFile = app.Flag("output-summary-file", "Path to save the summary of packages downloaded").String()
+	repoUrls          = app.Flag("repo-url", "URLs of the repos to download from.").Strings()
+	workerTar         = app.Flag("worker-tar", "Full path to worker_chroot.tar.gz").Required().ExistingFile()
+	buildDir          = app.Flag("worker-dir", "Directory to store chroot while running repo query.").String()
+
+	concurrentNetOps = app.Flag("concurrent-net-ops", "Number of concurrent network operations to perform.").Default(defaultNetOpsCount).Uint()
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	prof, err := profile.StartProfiling(profFlags)
+	if err != nil {
+		logger.Log.Warnf("Could not start profiling: %s", err)
+	}
+	defer prof.StopProfiler()
+
+	timestamp.BeginTiming("precacher", *timestampFile)
+	defer timestamp.CompleteTiming()
+
+	// Get the state of the current repo from a snapshot
+	rpmSnapshot, err := rpmSnapshotFromFile(*snapshot)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	// See what is available online from the upstream repos
+	availablePackages, err := getAllRepoData(*repoUrls, *workerTar, *buildDir)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Found %d available packages", len(availablePackages))
+	if logger.Log.IsLevelEnabled(logrus.DebugLevel) {
+		for _, pkg := range availablePackages {
+			logger.Log.Debugf("Found package: %s", pkg)
+		}
+	}
+
+	// For each package in the snapshot, check if it is available online and try to download it
+	downloadedPackages, err := downloadMissingPackages(rpmSnapshot, availablePackages, *outDir, *concurrentNetOps)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Downloaded %d packages into the cache", len(downloadedPackages))
+	err = writeSummaryFile(*outputSummaryFile, downloadedPackages)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+}
+
+func rpmSnapshotFromFile(snapshotFile string) (rpmSnapshot *repocloner.RepoContents, err error) {
+	err = jsonutils.ReadJSONFile(snapshotFile, &rpmSnapshot)
+	return
+}
+
+// getAllRepoData returns a map of package names to URLs for all packages available in the given repos. It uses
+// a chroot to run repoquery.
+func getAllRepoData(repoUrls []string, workerTar, buildDir string) (namesToUrls map[string]string, err error) {
+	const (
+		leaveChrootOnDisk = false
+	)
+	timestamp.StartEvent("pull available package data from repos", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Create a chroot to run repoquery in
+	// Create the directory 'buildDir' if it does not exist
+	exists, err := file.DirExists(buildDir)
+	if err != nil {
+		err = fmt.Errorf("failed to check if directory %s exists: %w", buildDir, err)
```

Comment:
Could we surround file/directory names with brackets or single quotes to make debugging easier? This helps a lot, if by mistake the variable ends up being an empty string.